tests/integration/utils/kafka_server.py

```diff
@@ -0,0 +1,164 @@
+            # - if the address/port does not point to a running server
+            # LeaderNotAvailableError:
+            # - if there is no leader yet
+            except (
+                NoBrokersAvailable,
+                LeaderNotAvailableError,
+                ValueError,
+            ) as e:
+                print(f"Error checking kafka cluster: {e}")
+                time.sleep(2.0)
+            else:
+                list_topics_successful = True
+
+
+def maybe_download_kafka(kafka_description: KafkaDescription) -> None:
+    """If necessary download kafka to run the tests."""
+    if not os.path.exists(kafka_description.install_dir):
+        logging.info("Downloading Kafka {url}", url=kafka_description.download_url)
+
+        download = requests.get(kafka_description.download_url, stream=True)
+        with tarfile.open(mode="r:gz", fileobj=download.raw) as file:
+            file.extractall(str(kafka_description.install_dir.parent))
+
+
+def kafka_java_args(
+    heap_mb: int,
+    kafka_config_path: str,
+    logs_dir: str,
+    log4j_properties_path: str,
+    kafka_description: KafkaDescription,
+) -> List[str]:
+    msg = f"Couldn't find kafka installation at {kafka_description.install_dir} to run integration tests."
+    assert kafka_description.install_dir.exists(), msg
+    java_args = [
+        "-Xmx{}M".format(heap_mb),
+        "-Xms{}M".format(heap_mb),
+        "-Dkafka.logs.dir={}/logs".format(logs_dir),
+        "-Dlog4j.configuration=file:{}".format(log4j_properties_path),
+        "-cp",
+        str(kafka_description.install_dir / "libs" / "*"),
+        "kafka.Kafka",
+        kafka_config_path,
+    ]
+    return java_args
+
+
+def configure_and_start_kafka(
+    zk_config: ZKConfig,
+    kafka_config: KafkaConfig,
+    kafka_description: KafkaDescription,
+    log4j_config: str,
+) -> Popen:
+    config_path = Path(kafka_config.logdir) / "server.properties"
+
+    advertised_listeners = ",".join(
+        [
+            f"PLAINTEXT://127.0.0.1:{kafka_config.plaintext_port}",
+        ]
+    )
+    listeners = ",".join(
+        [
+            f"PLAINTEXT://:{kafka_config.plaintext_port}",
+        ]
+    )
+
+    # Keep in sync with containers/docker-compose.yml
+    kafka_ini = {
+        "broker.id": 1,
+        "broker.rack": "local",
+        "advertised.listeners": advertised_listeners,
+        "auto.create.topics.enable": False,
+        "default.replication.factor": 1,
+        "delete.topic.enable": "true",
+        "inter.broker.listener.name": "PLAINTEXT",
+        "inter.broker.protocol.version": kafka_description.protocol_version,
+        "listeners": listeners,
+        "log.cleaner.enable": "true",
+        "log.dirs": kafka_config.datadir,
+        "log.message.format.version": kafka_description.protocol_version,
+        "log.retention.check.interval.ms": 300000,
+        "log.segment.bytes": 200 * 1024 * 1024,  # 200 MiB
+        "log.preallocate": False,
```

Comment:
I'm leaving this behind, even though it may not be necessary now that logs and data are on different directories, but I added this because on the first try everything was included and the generated artifact was a couple of GiB because the files had been preallocated.